src/modules/types.rs

```diff
@@ -1,26 +1,64 @@
 use std::fmt::Display;
 
 use heraclitus_compiler::prelude::*;
+use itertools::Itertools;
 use crate::utils::ParserMetadata;
 
-#[derive(Debug, Clone, PartialEq, Eq, Default)]
+#[derive(Debug, Clone, Eq, Default)]
 pub enum Type {
     #[default] Null,
     Text,
     Bool,
     Num,
+    Union(Vec<Type>),
     Array(Box<Type>),
     Failable(Box<Type>),
     Generic
 }
 
+impl Type {
+    fn eq_union_normal(one: &[Type], other: &Type) -> bool {
+        one.iter().any(|x| (*x).to_string() == other.to_string())
+    }
+
+    fn eq_unions(one: &[Type], other: &[Type]) -> bool {
+        let bigger;
+        let smaller = if one.len() < other.len() { bigger = other; one } else { bigger = one; other };
+
+        smaller.iter().all(|x| {
+            Self::eq_union_normal(bigger, x)
+        })
+    }
+}
+
+impl PartialEq for Type {
+    fn eq(&self, other: &Self) -> bool {
+        match (self, other) {
+            (Type::Null, Type::Null) => true,
+            (Type::Text, Type::Text) => true,
+            (Type::Bool, Type::Bool) => true,
+            (Type::Num, Type::Num) => true,
+            (Type::Generic, Type::Generic) => true,
+
+            (Type::Array(ref a), Type::Array(ref b)) => a == b,
+            (Type::Failable(ref a), Type::Failable(ref b)) => a == b,
+            
+            (Type::Union(one), Type::Union(other)) => Type::eq_unions(one, other),
+            (Type::Union(one), other) => Type::eq_union_normal(one, other),
+            (other, Type::Union(one)) => Type::eq_union_normal(one, other),
```

Comment:
Can we perhaps implement `PartialEq` for just the `Union` struct? I agree with @mks-h that we shouldn't make the same mistake with defining that custom `PartialEq`

Reply:
yeah, that thought crossed my mind. i will expirement around that today